src/mpid/common/thread/mpidu_thread_fallback.h

```diff
@@ -95,8 +95,62 @@ typedef MPL_thread_func_t MPIDU_Thread_func_t;
 - _context - A context (typically an object) of the critical section
 
 M*/
+
+/*M MPIDU_THREAD_CS_TRYENTER - Try enter a named critical section
+
+  Input Parameters:
++ _name - name of the critical section
+- _context - A context (typically an object) of the critical section
+- cs_acqt  - A flag that indicades whether the critical section was acquired
+
+M*/
+
+/*M MPIDU_THREAD_CS_EXIT - Exit a named critical section
+
+  Input Parameters:
++ _name - cname of the critical section
+- _context - A context (typically an object) of the critical section
+
+M*/
+
+/*M MPIDU_THREAD_CS_YIELD - Temporarily release a critical section and yield
+    to other threads
+
+  Input Parameters:
++ _name - cname of the critical section
+- _context - A context (typically an object) of the critical section
+
+  M*/
+
+/*M MPIDU_THREAD_CS_ENTER_ST - Stateful version of CS_ENTER
+ * It takes a MPIDU_thread_state_t variable to pass the state between ENTER and EXIT.
+  M*/
+
+/*M MPIDU_THREAD_CS_EXIT - Stateful version of CS_EXIT
+ * It takes a MPIDU_thread_state_t variable to pass the state between ENTER and EXIT.
+  M*/
+
+/* NOTE: this is currently only used by ch3:sock and only with GRANULARITY_GLOBAL */
+
+#if defined(MPICH_IS_THREADED)
 #define MPIDU_THREAD_CS_ENTER(name, mutex) MPIDUI_THREAD_CS_ENTER_##name(mutex)
+#define MPIDU_THREAD_CS_TRYENTER(name, mutex, cs_acq) MPIDUI_THREAD_CS_TRYENTER_##name(mutex, cs_acq)
+#define MPIDU_THREAD_CS_EXIT(name, mutex) MPIDUI_THREAD_CS_EXIT_##name(mutex)
+#define MPIDU_THREAD_CS_YIELD(name, mutex) MPIDUI_THREAD_CS_YIELD_##name(mutex)
+#define MPIDU_THREAD_CS_ENTER_ST(name, mutex, st) MPIDUI_THREAD_CS_ENTER_ST_##name(mutex,st)
+#define MPIDU_THREAD_CS_EXIT_ST(name, mutex, st) MPIDUI_THREAD_CS_EXIT_ST_##name(mutex,st)
+
+#else
+#define MPIDU_THREAD_CS_ENTER(name, mutex)      /* NOOP */
```

Comment:
`do {} while(0)` is a more robust way to implement the noop. Without the `do {} while (0)`, errors could successfully compile incorrectly. For example, the compiler wouldn't catch the error is we build the following with the `per-vci` critical section.
```
trying_out_new_code // forgot to terminate with a semicolon
CS_ENTER_GLOBAL();
```

Reply:
I agree that `do-while` is more robust for macros, but I find it less clear to read.